### PR TITLE
[autograd] Defer device thread creation when multithreading is disabled

### LINKED ISSUE #184783: `set_multithreading_enabled` still spawns autograd threads
### 🐛 Describe the bug

The documentation for [`set_multithreading_enabled`](https://docs.pytorch.org/docs/2.12/generated/torch.autograd.grad_mode.set_multithreading_enabled.html) says:

> Ordinarily, when [accelerator](https://docs.pytorch.org/docs/2.12/torch.html#accelerators) devices are in use, the backward pass runs on device-specific worker threads. The engine creates these threads based on the number of available devices and reuses them across iterations.
> When mode=False, the backward pass runs on the calling thread instead

However, it seems to still spawn threads

```python
import os, torch

def autograd_threads():
    """Return list of native thread names containing 'autograd' (Linux)."""
    names = []
    for tid in os.listdir("/proc/self/task"):
        try:
            name = open(f"/proc/self/task/{tid}/comm").read().strip()
            if "autograd" in name:
                names.append(name)
        except OSError:
            pass
    return names

torch.autograd.grad_mode.set_multithreading_enabled(False)
print(f"autograd threads before backward: {autograd_threads()}")
x = torch.ones(100, requires_grad=True)
x.mean().backward()
threads = autograd_threads()
print(f"autograd threads after backward:  {threads}")
```

```python
autograd threads before backward: []
autograd threads after backward:  ['pt_autograd_0', 'pt_autograd_1', 'pt_autograd_2', 'pt_autograd_3', 'pt_autograd_4', 'pt_autograd_5', 'pt_autograd_6', 'pt_autograd_7']
```

I feel like these threads should not be spawned at all in https://github.com/pytorch/pytorch/blob/becedcb06e19598cfd1fd82f887dea7044c753ed/torch/csrc/autograd/engine.cpp#L1422

Although then they would need to be lazy initialized later on if the context manager was switched back to mode=True later on.

Is this unexpected or is this a limitation of `set_multithreading_enabled` and the documentation should mention it?

Thank you

### Versions

2.11.0

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### PR DESCRIPTION
**Pull request by XAheli**
## Summary

Fixes #184783

**Root cause:** `Engine::initialize_device_threads_pool()` used a function-local `static bool` lambda that unconditionally called `start_device_threads()` on the very first backward — without checking `get_multithreading_enabled()`. Additionally, `start_device_threads()` bundled two unrelated concerns: allocating `thread_pool_shared_` (needed always for re-entrant backward) and spawning per-device worker threads (only needed when multithreading is enabled).

**Fix:**

- Move `thread_pool_shared_` initialization to the `Engine` constructor so it is always available for re-entrant backward regardless of multithreading state
- Early-return from `initialize_device_threads_pool()` when `get_multithreading_enabled()` is false, skipping device thread creation entirely
- Replace the function-local `static bool` with `c10::call_once` / `c10::once_flag` as a member variable for lazy, thread-safe, exception-safe device thread creation on the first backward where multithreading is actually enabled
- As a member variable (not function-local static), `device_threads_flag_` correctly resets on engine reconstruction after fork via `PythonEngine::get_python_engine()`

## Test plan

5 new subprocess-based tests in `TestMultithreadAutograd` (CUDA + Linux):

- `test_no_device_threads_when_multithreading_disabled` — core bug fix verification
- `test_device_threads_created_on_reenable` — lazy creation after re-enabling multithreading
- `test_device_threads_default_behavior` — regression guard for default (enabled) behavior
- `test_multithreading_disabled_cuda_correctness` — gradient numerical correctness
- `test_context_manager_deferred_threads` — context manager lifecycle

```bash
python test/test_autograd.py TestMultithreadAutograd -v                        # 17/17 pass
python test/test_autograd.py TestAutograd -v                                   # 388/388 pass
python test/test_autograd.py TestAutogradDeviceTypeCUDA -v                      # 67/67 pass
python test/test_autograd.py TestAutogradMultipleDispatchCUDA -v                # 12/12 pass
python test/test_autograd.py TestAutogradStreamSynchronizationCUDA -v           # 7/7 pass
```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### CODE REVIEW

**Comment by ezyang** [comment]
@albanD this looks OK. Did you want to look too?

**Comment by ezyang** [comment]
@claude Exhaustively analyze all initialization paths, as well as restart engine and fork paths, and check if this is correct. I am concerned about creating thread pool earlier; it seems like it should still be safe to create it later.

**Comment by XAheli** [comment]
Is this ready to merge?

@ezyang @albanD

**Comment by ezyang** [comment]
@claude Analyze for potential deadlocks

**Comment by ezyang** [comment]
@claude Alban is specifically worried about this:

> If your first backward is with multithreading disabled and the next enabled, I guess you're going to smoothly deadlock? 😛

**Review by albanD — COMMENTED** [review]
Thanks for sending a PR. That sounds ok for the worker threads but it would be great to have a test for the reentrant thread pool as well!

**Comment by ezyang** [comment]
@albanD I'm deferring to you

**Comment by XAheli** [comment]
Hi @albanD Lmk if anything else is needed or if this looks good to land!

**Comment by XAheli** [comment]
cc @groenenboomj

File: test/test_autograd.py [hunk 1]
```diff
@@ -15758,6 +15758,282 @@ def test_remove_obj_from_tls(self):
...
+            text=True,
+        )
+        self.assertEqual(
+            result.returncode,
+            0,
+            f"Default behavior test failed: {result.stderr}",
+        )
+
+    @unittest.skipIf(not TEST_CUDA, "test requires CUDA")
+    def test_multithreading_disabled_cuda_correctness(self):
```
Comment (albanD) [inline test/test_autograd.py]: What is this testing?

File: test/test_autograd.py [hunk 2]
```diff
@@ -15758,6 +15758,282 @@ def test_remove_obj_from_tls(self):
         torch._C._remove_obj_from_tls("test_obj")
         self.assertFalse(torch._C._is_key_in_tls("test_obj"))
 
+    @unittest.skipIf(not IS_LINUX, "requires /proc filesystem")
```
Comment (albanD) [inline test/test_autograd.py:16159]: Are there any non-linux tests here? We should make sure it works on macos/windows as well.

File: test/test_autograd.py [hunk 3]
```diff
@@ -15758,6 +15758,282 @@ def test_remove_obj_from_tls(self):
...
+    @unittest.skipIf(not TEST_CUDA, "test requires CUDA")
```
Comment (albanD) [inline test/test_autograd.py]: All of these should be accelerator, not cuda-specific

File: test/test_autograd.py [hunk 4]
```diff
@@ -15758,6 +15758,282 @@ def test_remove_obj_from_tls(self):
...
+        # We run in a subprocess to ensure a fresh engine state.
+        script = """\
+import os, sys, torch
+
+torch.autograd.set_multithreading_enabled(False)
+x = torch.ones(100, device="cuda", requires_grad=True)
+(x * 2).sum().backward()
+
+threads = []
+for tid in os.listdir("/proc/self/task"):
```
Comment (albanD) [inline test/test_autograd.py]: Isn't most of this code re-used below? Can we refactor this to reduce duplicate code?

File: test/test_autograd.py (hunk 1 above)
Comment (XAheli) [reply test/test_autograd.py]: To verify gradient numerical correctness with multithreading disabled. Splitted into two tests with clearer names.

File: test/test_autograd.py (hunk 2 above)
Comment (XAheli) [reply test/test_autograd.py:16159]: addressed

File: test/test_autograd.py (hunk 3 above)
Comment (XAheli) [reply test/test_autograd.py]: addressed

File: test/test_autograd.py (hunk 4 above)
Comment (XAheli) [reply test/test_autograd.py]: addressed

**Comment by XAheli** [comment]
FYI @albanD I've addressed the review comments. Could you PTAL when feasible :)
